workflow/src/main/java/bio/terra/tanagra/workflow/GraphUtils.java

```diff
@@ -0,0 +1,131 @@
+package bio.terra.tanagra.workflow;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Lists;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.Flatten;
+import org.apache.beam.sdk.transforms.KvSwap;
+import org.apache.beam.sdk.transforms.ParDo;
+import org.apache.beam.sdk.transforms.join.CoGbkResult;
+import org.apache.beam.sdk.transforms.join.CoGroupByKey;
+import org.apache.beam.sdk.transforms.join.KeyedPCollectionTuple;
+import org.apache.beam.sdk.values.KV;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.PCollectionList;
+import org.apache.beam.sdk.values.TupleTag;
+
+/** Utilities for working with graphs in Apache Beam. */
+public final class GraphUtils {
+  private GraphUtils() {}
+
+  /**
+   * Computes the transitive closure of a directed acyclic graph defined by {@code edges}.
```

Comment:
Out of curiosity, what happens if this is not acyclic? Is that actually a prerequisite of this function?

Reply:
Nope, though it will generate a lot of duplicates. Might want to add calls to `Distinct` to deduplicate if that happens often.